docs/design-proposals/2023-05 vParquet3.md

```diff
@@ -0,0 +1,179 @@
+  * (Stretch) Find attributes that are search the most by
+
+Future iterations may support additional features:
+
+* Workload analysis to determine suitable dedicated column configurations which can be used to make suggestions or to automate the configuration
+* Fully dynamic schema that allows to add dedicated columns for different data types and encodings (e.g. delta or run length encoding)
+* Apply config dynamically to old blocks during compaction
+
+## Implementation
+
+### Schema with spare columns
+
+There are two approaches for storing span attributes in dedicated columns in Tempo.
+One possibility is to extend the Parquet schema dynamically and add the required extra columns at runtime.
+This approach would allow to configure the number, type, and encoding of dedicated attribute columns and therefore allow Tempo to optimize the block layout specifically for every use case.
+Ultimately this is what Tempo should be capable of in the future.
+However, handling dynamically changing Parquet schemata correctly adds a lot of complexity.
+
+The alternative approach is to keep a set of spare columns in the schema and dynamically assign them to the respective span attributes.
+The downside here is that the number, type, and encoding for dedicated span attributes is very restricted.
+To keep the added complexity manageable, dedicated columns for span attributes are implemented using the spare columns approach for now.
+
+In the analyzed use cases, the 20 span attributes that contributed most to the overall attribute size,
+were almost exclusively attributes with string values.
+Given that the benchmark achieved good results with 5 dedicated columns,
+it can be expected that extending the schema by 10 spare columns for string values is sufficient to achieve good performance improvements for most use cases.
+
+The following code outlines the schema changes:
+
+```go
+type DedicatedAttrs struct {
+    String01 *string `parquet:",snappy,optional,dict"`
+    String02 *string `parquet:",snappy,optional,dict"`
+    ...
+    String10 *string `parquet:",snappy,optional,dict"`
+}
+
+type Span struct {
+    ...
+    DedicatedAttrs DedicatedAttrs `parquet:""`
+}
+```
+
+To execute TraceQL queries it is necessary to know which attribute is assigned to which spare column.
+Since the assignment of spare columns to specific attributes changes over time, it is necessary to store this information for each block.
+
+Tempo stores metadata about each Block in the `meta.json` file.
+The schema of this file will be extended such that it contains the assignment information:
+
+```go
+{
+    ...
+    "dedicatedColumns": [
+        {"scope": "span", "name": "attrA", "type": "string"},
+        {"scope": "span", "name": "attrB", "type": "string"},
+        {"scope": "span", "name": "attrC", "type": "string"},
+    ]
+}
+```
+
+## Configuration and overrides
+
+Dedicated columns have been part of the Parquet-based formats since the first version.
```

Comment:
this raises the question: are we planning on removing the old dedicated columns in favor of this new model?

Reply:
In future iterations, likely yes. For now, this new format is only an extension over the current schema, so they stay.

Reply:
I'm fine with this, but would love to see the old special columns removed. It will clean up a lot of code.